### PR TITLE
Improved content administration (image & category)

### PR DESCRIPTION
**Pull request by Elfangor93**
This pull request improves the administration of galleries - especially of those with a huge amount of categories and images. It adds some handy features to the category as well as the image manager in order to manage the content (images and categories) easier.
The following improvements were made:

**Category Manager**
- Category filter (only selected category and its child categories are shown)
- Level filter (only subcategories down to a selected level deepness is shown. Level zero is the root category or the category chosen in the category filter)
- Images row (additional row which shows the number of images in this category. Link to the images of that category in the image manager)
- Ordering row (buttons to move up or down a category by one position. Less fiddling than the current dragndrop positioning of the categories.) --> Has to be activated in the config manager
- Alias regeneration in batch edit mode of categories (regeneration of alias of the selected categories based on its current category structure)

**Image Manager**
- Category row (link to the subcategories of that category in the category manager)
- Ability to show and edit additional fields from plugins in the batch image form (see [PR #1](https://github.com/JoomGalleryfriends/JoomAdditionalImageFields/pull/1) on JoomAdditionalImageFields Plugin)
- Add a search and replace button to the toolbar (perform a search and replace on the selected image fields of the selected image)
- Ability to perform search and replace on additional fields from plugins (new plugin event called onJoomAfterSearchReplace)

### CODE REVIEW

**Comment by Elfangor93** [comment]
There is a small bug in this PR which I have no idea how to solve it...

**Steps to reproduce the bug:**
1. Open category manager
2. Sort categories by ordering asc
3. Move one category using the three dots on the left side of the row

**Expected behavior:**
Buttons to move up and down the category order position gets updated according to the new category order.

**Actual behavior:**
Buttons dot gets updated and showing actions which can not be performed anymore.

The problem is that the moving with the tree dots is done in using ajax request. But the buttons only gets updated when reloading the page.

Anyone an idea how to solve tis issue?

**Comment by MrMusic** [comment]
Thanks for this suggestion. In general, I like the improvements a lot. 
I haven't tested everything yet, but I have noticed a few points:

Category Manager:
Sorting the categories with the buttons: I'm not sure if we should do that. This sort of sorting already existed in Joomla 2.5. ;-)
With Joomla 3, sorting by drag and drop was introduced. The JoomGallery has adapted to Joomla here. Also with Joomla 4 there is still only the sorting with drag and drop.
Are there any other opinions about this?

Regenerate alias
I think a description should be added here what the 'regnerate alias' means exactly.

Image Manager:
I would be in favour of integrating the 'search and replace' into the already existing batch form instead of opening an additional form.

**Comment by Elfangor93** [comment]
`Sorting the categories with the buttons: I'm not sure if we should do that. This sort of sorting already existed in Joomla 2.5. ;-)`

Sorting Categories with Drag'n'Drop gets very slushy when you have a lot of cats and subcats. When you grab a row all rows with another layer gets hided. This jumping around of cats confuses and makes it hard to place the current grabbed row correctly.

Additionally the Drag'n'Drop rearranging does work only 95% of the time. I dont know why but sometimes you rearrange and after reload the arrangement is again as before rearranging. I have never figured out why but this doesnt happen with the buttons.

I wouldnt drop a good working GUI element just because Joomla core did.

**Comment by Elfangor93** [comment]
`I would be in favour of integrating the 'search and replace' into the already existing batch form instead of opening an additional form.`

No, I dont want the search'n'replase inside the batch form. I would agree to adding another button to the toolbar at the top for the search'n'replase feature. But I think the toolbar of the images view is already crowded.

**Comment by Elfangor93** [comment]
There is now a separate button in the images toolbar for the search and replace feature.

**Comment by Elfangor93** [comment]
> 
> 
> There is a small bug in this PR which I have no idea how to solve it...
> 
> **Steps to reproduce the bug:**
> 
>     1. Open category manager
> 
>     2. Sort categories by ordering asc
> 
>     3. Move one category using the three dots on the left side of the row
> 
> 
> **Expected behavior:** Buttons to move up and down the category order position gets updated according to the new category order.
> 
> **Actual behavior:** Buttons dot gets updated and showing actions which can not be performed anymore.
> 
> The problem is that the moving with the tree dots is done in using ajax request. But the buttons only gets updated when reloading the page.
> 
> Anyone an idea how to solve tis issue?

I propose to add a new configuration parameter where you can choose if you want to have to drag n drop sorting or sorting with the buttons. This would solve the bug mentioned above and reduces confusion for users who are used to the drag n drop sorting.
The default for the config parameter would be dragndrop...

**Comment by MrMusic** [comment]
Thanks for changing this. It looks better for me.
While testing, I noticed the following:
In images view the sorting via buttons is not possible, is this intended?
In categories view the 'search and replace' is not possible, is this intended?

When search and replace, currently only terms consisting of one (1) word are taken into account. This should be pointed out in a description.
Maybe use the filter 'string' instead of 'cmd' to make replacements with several words possible?

In filter_categories.xml the `<option value="0">JNONE</option>` option is useful for?
I would change `<option value="-1">JOPTION_SELECT_MAX_LEVELS</option>` to `<option value="">JOPTION_SELECT_MAX_LEVELS</option>` and remove the `<option value="0">JNONE</option>` 
also to prevent the undesired display of the 'Search Tools'.

**Comment by Elfangor93** [comment]
> In images view the sorting via buttons is not possible, is this intended?
> In categories view the 'search and replace' is not possible, is this intended?
Yes, you can see in my first comment what I have added to which list view. Sorting with buttons, s'n'r and additional fields in batch form for category manager is not included in this PR.

Right now I dont need this functions in any of my projects. Could be something for a future PR (you could open a bounty for it...)

**Comment by MrMusic** [comment]
>Sorting with buttons, s'n'r and additional fields in batch form for category manager is not included in this PR

Ok, but the description says that it should also work for the categories: _Select a method to sort elements (images, categories)..._

**Comment by Elfangor93** [comment]
> Ok, but the description says that it should also work for the categories

Now it should be clear...

File: administrator/components/com_joomgallery/views/images/tmpl/default_replace_body.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        <div class="row-fluid">
+          <div class="span3">
+            <label class="checkbox"><input type="checkbox" name="batch_fields[]" value="imgtitle"> <?php echo JText::_('COM_JOOMGALLERY_COMMON_TITLE'); ?></label>
+            <label class="checkbox"><input type="checkbox" name="batch_fields[]" value="imgtext"> <?php echo JText::_('COM_JOOMGALLERY_COMMON_DESCRIPTION'); ?></label>
+            <label class="checkbox"><input type="checkbox" name="batch_fields[]" value="imgauthor"> <?php echo JText::_('COM_JOOMGALLERY_COMMON_AUTHOR'); ?></label>
+          </div>
+          <div class="span3">
+            <label class="checkbox"><input type="checkbox" name="batch_fields[]" value="imgtitle"> <?php echo JText::_('COM_JOOMGALLERY_COMMON_METADESC_TIP'); ?></label>
+            <label class="checkbox"><input type="checkbox" name="batch_fields[]" value="imgtext"> <?php echo JText::_('COM_JOOMGALLERY_COMMON_METAKEYS_TIP'); ?></label>
+            <label class="checkbox"><input type="checkbox" name="batch_fields[]" value="additional"> <?php echo JText::_('COM_JOOMGALLERY_COMMON_ADDITIONALIELDS'); ?></label>
```
[the remark this answers is not in the document]
Comment (Elfangor93) [reply administrator/components/com_joomgallery/views/images/tmpl/default_replace_body.php]: Fixed with commit a6a7b6e

**Comment by MrMusic** [comment]
Ordering in Category Manager:
Changing the sorting from ordering ascending to ordering descending has no effect.
I think the ordering buttons should only be enabled if ordering ascending is activated.

Search and Replace:
Since the search is 'case sensitive', this should be pointed out in a description.

**Comment by Elfangor93** [comment]
Thank you @MrMusic. Your comments have been implemented.

**Comment by MrMusic** [comment]
I have tested this item ✅ successfully. Thank you very much.

**Pull request merged**